### PR TITLE
Use the current parser location for CSS error

### PR DESCRIPTION
**Pull request by SimonSapin**
… rather than the start location of the current construct. This likely places the error just *after* of the unexpected token whereas before would be best, but that’s likely a much bigger change.

See https://bugzilla.mozilla.org/show_bug.cgi?id=1378861

---
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/18585)

### CODE REVIEW

**Comment by SimonSapin** [comment]
**Do not merge yet**, waiting on Gecko try results to have a patch ready to update tests.

@bors-servo try

**Comment by jdm** [comment]
@SimonSapin Needs a rebase.

**Comment by SimonSapin** [comment]
Thanks Josh.

@bors-servo try

**Comment by SimonSapin** [comment]
``error: Could not compile `script`.`` (I think from Cargo) with no error message from rustc, and I can’t reproduce locally

@bors-servo retry

**Comment by SimonSapin** [comment]
Gecko try: https://treeherder.mozilla.org/#/jobs?repo=try&revision=a4a1069a44de1930da3bb086529f2e705810cdbe

This depends on https://github.com/servo/rust-cssparser/pull/200 but otherwise should be ready to land, I think. r? @emilio

File: components/selectors/parser.rs [hunk 1]
```diff
@@ -1050,24 +1047,28 @@ fn display_to_css_identifier<T: Display, W: fmt::Write>(x: &T, dest: &mut W) ->
...
         // Parse a sequence of simple selectors.
-        parsed_pseudo_element = match parse_compound_selector(parser, input, &mut builder) {
-            Ok(result) => result,
-            Err(ParseError::Custom(SelectorParseError::EmptySelector)) if builder.has_combinators() =>
-                return Err(SelectorParseError::DanglingCombinator.into()),
-            Err(e) => return Err(e),
+        has_pseudo_element = match parse_compound_selector(parser, input, &mut builder)? {
+            Some(has_pseudo_element) => has_pseudo_element,
+            None => {
+                if builder.has_combinators() {
```
Comment (emilio) [inline components/selectors/parser.rs]: nit: You can use `return if` if you want here.

File: components/selectors/parser.rs [hunk 2]
```diff
@@ -1228,16 +1228,21 @@ fn parse_qualified_name<'i, 't, P, E, Impl>
...
             Ok(&Token::Delim('*')) if !in_attr_selector => {
                 Ok(OptionalQName::Some(namespace, None))
             },
             Ok(&Token::Ident(ref local_name)) => {
                 Ok(OptionalQName::Some(namespace, Some(local_name.clone())))
             },
-            Ok(t) if in_attr_selector => Err(SelectorParseError::InvalidQualNameInAttr(t.clone()).into()),
-            Ok(t) => Err(SelectorParseError::ExplicitNamespaceUnexpectedToken(t.clone()).into()),
-            Err(e) => Err(ParseError::Basic(e)),
+            Ok(t) if in_attr_selector => Err(location.new_custom_error(
```
Comment (emilio) [inline components/selectors/parser.rs]: nit: Use a block for multi-line arms?

File: components/selectors/parser.rs [hunk 3]
```diff
@@ -1299,16 +1306,18 @@ fn parse_qualified_name<'i, 't, P, E, Impl>
 }
 
 
-fn parse_attribute_selector<'i, 't, P, E, Impl>(parser: &P, input: &mut CssParser<'i, 't>)
-                                                -> Result<Component<Impl>,
-                                                          ParseError<'i, SelectorParseError<'i, E>>>
-    where P: Parser<'i, Impl=Impl, Error=E>, Impl: SelectorImpl
+fn parse_attribute_selector<'i, 't, P, Impl>(parser: &P, input: &mut CssParser<'i, 't>)
```
Comment (emilio) [inline components/selectors/parser.rs]: Feel free to reformat the signatures to use block indentation for arguments if you feel like it while touching them.

**Review by emilio — APPROVED** [review]
rs=me

**Comment by SimonSapin** [comment]
https://treeherder.mozilla.org/#/jobs?repo=try&revision=a4a1069a44de1930da3bb086529f2e705810cdbe look green enough.

@bors-servo r=emilio

Thanks for the review!

**Comment by SimonSapin** [comment]
Homu seems confused, I pushed a dummy amend…

@bors-servo r=emilio

**Comment by SimonSapin** [comment]
Bare ``error: Could not compile `script`.``, I’ve filed #18806.

@bors-servo retry

**Comment by SimonSapin** [comment]
Pushed another dummy amend to hopefully get homu unstuck.

@bors-servo r=emilio

**Comment by SimonSapin** [comment]
Moved to #18808.

**Pull request closed without merging**